### PR TITLE
refactor: rewrite YAML-generation pipeline

### PR DESCRIPTION
**Pull request by hegemonic**
This PR is based on discussions with @tbpg and @dansaadati. It reflects the following goals:

+ Structure the code more clearly, so that it's easier to add features and fix bugs.
+ Use JavaScript, not shell.
+ During post-processing, treat YAML files as objects, rather than processing them line-by-line.
+ During post-processing, read, process, and write each YAML file only once.
+ Provide full test coverage.

My intent for this PR was to avoid making any meaningful changes to the YAML output. I think I've achieved that goal, based on eyeballing some diffs between the previous version's output and this version's output. Let me know if I should do more.

@dandhlee FYI

---

Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com//issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea [_discussed offline with @tbpg and @dansaadati_]
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by hegemonic** [comment]
Working on the CI failure, FYI.

File: __snapshots__/generate-devsite.mjs.js
Comment (dandhlee) [inline __snapshots__/generate-devsite.mjs.js:211]: If we're separating these lines into actual YAML files, the `YamlMime` line is likely needed for the DocFX YAML converter to correctly translate them into corresponding HTML using the correct template (there's Universal as well as Managed types).

File: __snapshots__/generate-devsite.mjs.js
Comment (dandhlee) [inline __snapshots__/generate-devsite.mjs.js:1726]: Some lines are surrounded with quotes, other are not (i.e. `uid` entries vs `type`, `summary`, etc). Is there a reason for some of them to be with quotes and others without? It could just be that I'm not familiar with Node's way of dealing with this.

**Review by dandhlee — COMMENTED** [review]
Thank you for working on this Jeff! I'm not a JS expert so I'll leave it to @dansaadati to help take a look.

File: __snapshots__/generate-devsite.mjs.js
Comment (hegemonic) [reply __snapshots__/generate-devsite.mjs.js]: Updated to add this line to all YAML files except `toc.yml`, which appears to match the behavior of the existing code.

File: __snapshots__/generate-devsite.mjs.js
Comment (hegemonic) [reply __snapshots__/generate-devsite.mjs.js:1726]: I think this is caused by a couple of things:

+ In many cases, YAML doesn't care whether or not you enclose a string in quotes.
+ We're now parsing and stringifying each YAML file with the [`js-yaml` package](https://github.com/nodeca/js-yaml), which might have different quoting rules than whatever library `api-documenter` is using.

If I'm correct, then these differences shouldn't affect the generated docs. But I might be wrong. If someone who's more familiar with the DocFX converter could use this PR to generate some YAML files, then run them through the rest of the doc-generation pipeline, that would be helpful!

**Review by dansaadati — COMMENTED** [review]
Thanks for the work on this!

File: generate-devsite.mjs
Comment (hegemonic) [reply generate-devsite.mjs:28]: Done.

File: .editorconfig
Comment (hegemonic) [reply .editorconfig:1]: Ugh, sorry about that! Added to `.gitignore` as well.

File: main.mjs
Comment (hegemonic) [reply main.mjs:1]: I'm fine with that, but this package claims to support Node.js >=12.0.0. I believe that ES modules were behind a flag until Node.js 13.2/14.0, [per MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules#javascript.statements.import).

Are you okay with bumping the minimum Node.js version to 14? If so, I'll do that and rename all the `.mjs` files.

File: test/helpers.mjs
Comment (hegemonic) [reply test/helpers.mjs:54]: Yikes! Hopefully switching to `git-subtree` in a future PR will speed things up. But for now, increased the timeout.

File: main.mjs
Comment (dansaadati) [reply main.mjs:1]: Ah, that's okay. Let's leave this as is for now.

**Review by dansaadati — APPROVED** [review]

File: __snapshots__/generate-devsite.mjs.js
Comment (dansaadati) [reply __snapshots__/generate-devsite.mjs.js:1726]: I ran this through docfx, and I see, with our template at least, the `name`, `fullName`, and `summary` keys are not affected by being wrapped by quotes or not.

**Review by dansaadati — APPROVED** [review]

**Comment by hegemonic** [comment]
Friendly ping. Anything I can do to help move this PR along?

**Comment by dandhlee** [comment]
@hegemonic OwlBot made some minor updates, can you do a quick sanity check to see if the changes made need to be reverted back?

**Comment by hegemonic** [comment]
The changes to `.eslintignore` and `.nycrc` are incorrect and should be reverted. The change to `.prettierrc.js` looks okay.

**Comment by dandhlee** [comment]
The changes need to be made on https://github.com/googleapis/nodejs-cloud-rad/blob/main/owlbot.py otherwise OwlBot will keep reverting them.

You can make changes like this: https://github.com/googleapis/python-storage/blob/a9d61e43ef9101f2910e410278fba66058d318a8/owlbot.py#L54-L57

So we'd want something like 
```
s.replace(
    ".eslintignore",
    "**/coverage",
    """__snapshots__
    .coverage"""
)
```
`"""` is required for multi-line changes. And similar change for `.nycrc`:
```
s.replace(
    ".nycrc",
    "karma.conf.js\",",
    """karma.conf.js\",
    \".prettierrc.js\","""
)
```

**Comment by hegemonic** [comment]
@dandhlee Thanks for the pointers! I made the change; let's see what OwlBot says.

**Comment by dandhlee** [comment]
Owlbot check is failing because two commits were pushed at once; I forgot how to resolve it but I think it is doable, I'll try to take a look by tomorrow.

File: owlbot.py
Comment (dandhlee) [inline owlbot.py:44]:
```suggestion
node.owlbot_main(
    templates_excludes=[
        "README.md",
        ".kokoro/**",
        ".github/workflows/**",
        ".trampolinerc",
        ".mocharc.js",
        ".github/release-trigger.yml",
        ".github/release-please.yml",
        ".github/sync-repo-settings.yaml",
    ]
)

s.replace(
    ".eslintignore",
    "**/coverage",
    """__snapshots__
.coverage""",
)

s.replace(
    ".nycrc",
    "karma.conf.js\",",
    """karma.conf.js",
    ".prettierrc.js",""",
)
```

File: owlbot.py
Comment (dandhlee) [inline owlbot.py:31]:
```suggestion
```

File: owlbot.py
Comment (dandhlee) [inline owlbot.py:44]:
```suggestion
s.replace(
    ".eslintignore",
    "\*\*/coverage",
    """__snapshots__
.coverage""",
)

s.replace(
    ".nycrc",
    "karma\.conf\.js\",",
    """karma.conf.js",
    ".prettierrc.js",""",
)
```

**Comment by dandhlee** [comment]
I think I finally got it to work, can you take a look to see if everything looks good now? @hegemonic

**Comment by hegemonic** [comment]
@dandhlee Thanks, looks good!

**Comment by hegemonic** [comment]
@dandhlee Anything else you need from me, or is this ready to merge?

**Comment by dansaadati** [comment]
> @dandhlee Anything else you need from me, or is this ready to merge?

This is ready to mergge!

**Comment by hegemonic** [comment]
Awesome! I don't have write access, so I'll need one of y'all to merge this.

**Comment by dansaadati** [comment]
> Awesome! I don't have write access, so I'll need one of y'all to merge this.

Done, thanks again for helping out with this refactor. It's gonna make it a lot easier to address the feedback we need to address!

**Pull request merged**
